Adding migrated checkov rules to AWS terraform checks

mertcancoskuner opened the pull request:
Migrated the Checkov rules listed at https://github.com/returntocorp/semgrep-rules/issues/1631 except some of them such as WAFHasLogs or WafHasAnyRules.

mertcancoskuner committed:
Adding migrated checkov rules to AWS terraform checks

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/returntocorp/semgrep-rules?pullRequest=1689) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/returntocorp/semgrep-rules?pullRequest=1689) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/returntocorp/semgrep-rules?pullRequest=1689) it.</sub>

minusworld commented:
@mertcancoskuner Wow, thanks!! I've approved for CI and will look over the files sometime in the next few weeks. Many of the approvers (including myself) are on holiday, so we'll get to it when we can. Thanks again!

clintgibler commented:
Yay, this is awesome, thanks @mertcancoskuner!!  🙌 🎉  As @minusworld said, many of us are on holiday, but we'll review this as soon as we can. Thanks again and happy holidays :D

minusworld commented:
@mertcancoskuner We're back! Looking over the PR, it looks like the test files are using different test annotations from what the CI here expects.

The CI here uses a special `> semgrep --test` flag which looks for `# ruleid: <id_here>` directly above the expected match lines in files with the same name.

I added this annotation to aws-elasticache-automatic-backup-not-enabled as an example. Would you mind updating the test files to match this format? I'd also be happy to take half of them myself -- let me know.

Thanks!

mertcancoskuner commented:
> @mertcancoskuner We're back! Looking over the PR, it looks like the test files are using different test annotations from what the CI here expects.
> 
> The CI here uses a special `> semgrep --test` flag which looks for `# ruleid: <id_here>` directly above the expected match lines in files with the same name.
> 
> I added this annotation to aws-elasticache-automatic-backup-not-enabled as an example. Would you mind updating the test files to match this format? I'd also be happy to take half of them myself -- let me know.
> 
> Thanks!

@minusworld gotcha. I'll update the others accordingly.

mertcancoskuner committed:
Adding migrated checkov rules to AWS terraform checks

mertcancoskuner committed:
Adding missing ruleids in tf files

mertcancoskuner committed:
Adding missing ruleids in tf files

minusworld committed:
Use network_interfaces block in aws_launch_template pattern for ec2-public-ip rule

minusworld commented on terraform/aws/security/aws-cloudtrail-encrypted-with-cmk.yaml:
```diff
@@ -0,0 +1,26 @@
+        kms_key_id = ...
+        ...
+      }
+  - pattern-regex: (^aws_kms_key\.(.*))
```
It looks like the intent of this regex is to filter out cases where `kms_key_id` is set to something with `aws_kms_key` as a prefix. Is that right?

In that case you'll have to group the `pattern-not-inside` and regex under a parent `patterns` clause so that they go together. In addition, this might be easier to do by capturing the right hand side of `kms_key_id` in a metavariable and using `metavariable-pattern` to test it.

mertcancoskuner replied:
That is right. Although, in a second thought, that might not be needed at all and can be scrapped altogether. What do you think?

minusworld committed:
Move test annotation line in aws-iam-admin-policy-ssoadmin.tf

minusworld commented on terraform/aws/security/aws-ssm-document-logging-issues.yaml:
```diff
@@ -0,0 +1,31 @@
+      }
+  - pattern: content = "$STATEMENT"
+  - metavariable-pattern:
+      metavariable: $STATEMENT
```
It looks like this might be missing the language field so that it scans JSON or YAML.

To do two languages with `metavariable-pattern`, you may have to break this up into two `patterns` clauses, one with `pattern: content="$STATEMENT"` and `metavariable-pattern` for JSON, and a second one with `pattern: content="$STATEMENT"` and `metavariable-pattern` for YAML.

mertcancoskuner replied:
Will change it accordingly

minusworld committed:
Move test annotations in aws-lambda-environment-credentials.tf

minusworld commented on terraform/aws/security/aws-dynamodb-table-unencrypted.tf:
```diff
@@ -0,0 +1,72 @@
+  }
+
+  server_side_encryption {
+      enabled = true
```
This one will still get flagged because it does not contain the `kms_key_arn = ...` inside. Do you want to flag this one, or does the pattern need amending?

mertcancoskuner replied:
Yes, since it has no cmk, we want to flag this one

minusworld committed:
Move test annotations in aws-kms-key-wildcard-principal.tf

minusworld committed:
Move test annotations in aws-glacier-vault-any-principal.tf

minusworld commented on terraform/aws/security/aws-elb-access-logs-not-enabled.yaml:
```diff
@@ -0,0 +1,40 @@
+rules:
+- id: aws-elb-access-logs-not-enabled
+  patterns:
+  - pattern-either:
```
I think this rule is missing clauses for the `aws_alb` resource in addition to `aws_lb`

mertcancoskuner replied:
Yes, `aws_alb` is missing and needs fixing

minusworld commented on terraform/aws/security/aws-codebuild-project-unencrypted.tf:
```diff
@@ -0,0 +1,58 @@
+  service_role = "aws_iam_role.example.arn"
+
+  artifacts {
+    type = "NO_ARTIFACTS"
```
Do you have a preference for whether this case with NO_ARTIFACTS should be flagged? Currently it is flagged, and in that case it'll need a test annotation

mertcancoskuner replied:
It may cause false positives (which is valid for all unknown cases). Do you think it's worth increasing coverage for the sake of having more false positives?

minusworld replied:
I think it depends on whether this is an issue worth reporting or not. However, if we're uncertain, I'd say have more false positives -- if it annoys someone more knowledgeable than us, they will make an issue about it ;)

minusworld committed:
Add additional block to pattern in aws-codebuild-project-artifacts-unencrypted.yaml

minusworld commented on terraform/aws/security/aws-elasticsearch-nodetonode-encryption.tf:
```diff
@@ -0,0 +1,73 @@
+  }
+}
+
+resource "aws_elasticsearch_domain" "node_to_node_encryption_enabled" {
```
Should this one be flagged? It's currently highlighted: https://semgrep.dev/s/5yqX

mertcancoskuner replied:
This shouldn't be flagged. Need to work on this to identify what is the issue

minusworld commented on terraform/aws/security/aws-transfer-server-is-public.yaml:
```diff
@@ -0,0 +1,27 @@
+        resource "aws_transfer_server" $ANYTHING {
+          ...
+        }
+    - pattern: |
```
Based on the test case, is this supposed to be `pattern-not-inside`?

mertcancoskuner replied:
That's true, it should be `pattern-not-inside` to look for `endpoint_type = "VPC"`. Will fix

minusworld committed:
Move test annotations in aws-iam-admin-policy.tf

minusworld commented on terraform/aws/best-practice/aws-s3-bucket-versioning-not-enabled.tf:
```diff
@@ -0,0 +1,108 @@
+}
+
+
+resource "aws_s3_bucket" "this" {
```
This one also gets flagged. Could try filtering it out with `enabled = var.$ENABLED`.

Sadly we don't have TF variable resolution yet... but maybe someday!

minusworld commented on terraform/aws/best-practice/missing-aws-lb-deletion-protection.yaml:
```diff
@@ -0,0 +1,22 @@
+          ...
+        }
+    - pattern-not-inside: |
+        resource "aws_lb" $ANYTHING {
```
This rule looks like it's missing the `aws_alb` variants

minusworld commented on terraform/aws/best-practice/missing-aws-cross-zone-lb.yaml:
```diff
@@ -0,0 +1,22 @@
+rules:
+- id: missing-aws-cross-zone-lb
+  patterns:
```
Missing `aws_alb` cases

minusworld commented on terraform/aws/best-practice/aws-s3-object-lock-not-enabled.tf:
```diff
@@ -0,0 +1,46 @@
+
+# unknown
+
+resource "aws_s3_bucket" "default" {
```
Do you have a preference for whether this one gets flagged? Currently, the rule does flag here; however, this would alert on many S3 configurations.

If you want to alert here, it'll need a test annotation

mertcancoskuner replied:
It may cause false positives (which is valid for all unknown cases). Do you think it's worth increasing coverage for the sake of having more false positives?

minusworld replied:
Same as above

minusworld commented on terraform/aws/best-practice/aws-elasticache-automatic-backup-not-enabled.tf:
```diff
@@ -0,0 +1,41 @@
+  num_cache_nodes      = 1
+  parameter_group_name = "default.redis6.x"
+
+  snapshot_retention_limit = 0
```
The associated rule will not flag this one since the rule only checks for the presence of `snapshot_retention_limit` set to anything. You can filter this out with another `pattern-not-inside` clause checking for `0`.

minusworld commented on terraform/aws/best-practice/aws-rds-iam-authentication-not-enabled.tf:
```diff
@@ -0,0 +1,69 @@
+
+# unknown
+
+resource "aws_db_instance" "mariadb" {
```
Would you prefer if this one gets flagged? Currently it will get flagged.

mertcancoskuner replied:
It may cause false positives (which is valid for all unknown cases). Do you think it's worth increasing coverage for the sake of having more false positives?

minusworld reviewed (changes requested):
I left a bunch of comments.
They can be summarized as:
- All the `-cmk` rules need an alternate implementation. I've asked @lejliu to help figure out how to write these.
- There are a handful of rules that match on `aws_lb` but not `aws_alb`.
- Several of the test files have an `# unknown` section at the end - we'll have to decide if we want to prefer flagging these resources or not. I think I left comments on each one with my point of view, so hopefully they're easy to find and discuss.

I fixed a few of the minor issues as I went, like test annotations on the wrong line. Hopefully that means the results of the test run only have issues we actually need to address.

Thanks again! Let us know how we can be of help. :D

minusworld committed:
Move stray test annotation in aws-glacier-vault-any-principal.tf

mertcancoskuner commented:
> I left a bunch of comments. They can be summarized as:
> 
> * All the `-cmk` rules need an alternate implementation. I've asked @lejliu to help figure out how to write these.
> * There are a handful of rules that match on `aws_lb` but not `aws_alb`.
> * Several of the test files have an `# unknown` section at the end - we'll have to decide if we want to prefer flagging these resources or not. I think I left comments on each one with my point of view, so hopefully they're easy to find and discuss.
> 
> I fixed a few of the minor issues as I went, like test annotations on the wrong line. Hopefully that means the results of the test run only have issues we actually need to address.
> 
> Thanks again! Let us know how we can be of help. :D

- We might scrap `aws_kms_key` as terraform might be linting these already in case of any non-compliant input which would solve `-cmk` implementation.
- I'll add `aws_alb` cases wherever applicable.
- I think having fewer false positives is better in terms of overall user experience which I also prefer but having those may increase the rule coverage. What do you think?

minusworld commented:
> * We can remove the `aws_kms_key` check as terraform might be linting these already in case of any non-compliant input which would solve `-cmk` implementation.
> * I'll add `aws_alb` cases wherever applicable.
> * I think having fewer false positives is better in terms of overall user experience which I also prefer but having those may increase the rule coverage. What do you think?

I'm sure there's a way to get the `-cmk` rules to work... but we can remove them for now and put them up in another PR.

Re: `aws_alb`: great! Thank you 🙏 

Fewer false positives generally makes end users happier, in our experience. I think that makes sense, contrary to my inline comments above. 😅  If we err on the side of fewer false positives all around, that's OK.

mertcancoskuner committed:
Removed kms_key_id match, include alb in lb checks

mertcancoskuner committed:
Merge branch 'migrated-checkov-rules' of https://github.com/mertcancoskuner/semgrep-rules into migrated-checkov-rules

mertcancoskuner commented:
> > * We can remove the `aws_kms_key` check as terraform might be linting these already in case of any non-compliant input which would solve `-cmk` implementation.
> > * I'll add `aws_alb` cases wherever applicable.
> > * I think having fewer false positives is better in terms of overall user experience which I also prefer but having those may increase the rule coverage. What do you think?
> 
> I'm sure there's a way to get the `-cmk` rules to work... but we can remove them for now and put them up in another PR.
> 
> Re: `aws_alb`: great! Thank you 🙏
> 
> Fewer false positives generally makes end users happier, in our experience. I think that makes sense, contrary to my inline comments above. 😅 If we err on the side of fewer false positives all around, that's OK.

Removed `kms_key_id` comparison, fixed lb checks and ignored the unknowns. As you mentioned, if someone sees there is a good fix for cmk checks or unknowns causing false positives after community usage we can revise those as we go. There are two points that I couldn't manage to pull off;

> To do two languages with metavariable-pattern, you may have to break this up into two patterns clauses, one with pattern: content="$STATEMENT" and metavariable-pattern for JSON, and a second one with pattern: content="$STATEMENT" and metavariable-pattern for YAML.

This one (`aws-ssm-document-logging-issues.yaml`) started to give `metavariable-pattern failed because $STATEMENT does not bind to a sub-program` error which I couldn't figure out why

> The associated rule will not flag this one since the rule only checks for the presence of snapshot_retention_limit set to anything. You can filter this out with another pattern-not-inside clause checking for 0.

The solution that we discussed didn't solve the problem. I've tried `metavariable-comparison` with `$COUNT > 0` which resulted in failure as well.

minusworld committed:
Filter unknown case

minusworld committed:
Move test annotation

minusworld committed:
Add test annotation

minusworld committed:
Use TF parser instead of regex

minusworld committed:
Use specific resources instead of generic metavariables

minusworld committed:
Add test annotation to unknown case. It's very similar, only using mariadb instead

minusworld committed:
Use specific resources instead of generic metavariables

minusworld committed:
Fixes

minusworld committed:
Fail unknown case for expediency

minusworld committed:
Filter unknown case

minusworld committed:
Rewrite aws-lambda-environment-unencrypted

minusworld committed:
Filter unknown case

minusworld committed:
Fix owasp

minusworld commented:
@mertcancoskuner I went through and tried to fix up the remaining rules. There's only one left now that is giving me a bit of trouble. I think it can be written, so I'll try to get to it early next week.

minusworld commented:
Turns out there's an issue matching YAML documents in HCL: https://github.com/returntocorp/semgrep/issues/4582

minusworld committed:
Set YAML versions of aws-ssm-document-logging-issues to TODO

minusworld committed:
Fix OWASP

minusworld committed:
C'mon CLA-bot
